service/matching/matchingEngine.go

```diff
@@ -836,16 +841,17 @@ func (e *matchingEngineImpl) getTask(
 	return tlMgr.GetTask(ctx, maxDispatchPerSecond)
 }
 
-func (e *matchingEngineImpl) unloadTaskList(id *taskListID) {
+func (e *matchingEngineImpl) unloadTaskList(tlMgr taskListManager) {
+	id := tlMgr.TaskListID()
 	e.taskListsLock.Lock()
-	tlMgr, ok := e.taskLists[*id]
-	if ok {
-		delete(e.taskLists, *id)
+	currentTlMgr, ok := e.taskLists[*id]
+	if !ok || tlMgr != currentTlMgr {
+		e.taskListsLock.Unlock()
+		return
 	}
+	delete(e.taskLists, *id)
 	e.taskListsLock.Unlock()
```

Comment:
maybe we can defer this too and only have a single instance of unlock?

Reply:
I think we want/need it to unlock before `Stop()` runs :\

in a more general sense though, yeah - I've been hoping we can come up with a better structure for these kinds of things.  Now that we have generics, it might even be palatable to use an IIFE...

Reply:
ah, good point